src/Processors/QueryPlan/Optimizations/removeRedundantDistinct.cpp

```diff
@@ -0,0 +1,303 @@
+#include <memory>
+#include <Interpreters/ActionsDAG.h>
+#include <Processors/QueryPlan/AggregatingStep.h>
+#include <Processors/QueryPlan/ArrayJoinStep.h>
+#include <Processors/QueryPlan/CubeStep.h>
+#include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FillingStep.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/IntersectOrExceptStep.h>
+#include <Processors/QueryPlan/JoinStep.h>
+#include <Processors/QueryPlan/LimitByStep.h>
+#include <Processors/QueryPlan/LimitStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/RollupStep.h>
+#include <Processors/QueryPlan/SortingStep.h>
+#include <Processors/QueryPlan/TotalsHavingStep.h>
+#include <Processors/QueryPlan/UnionStep.h>
+#include <Processors/QueryPlan/WindowStep.h>
+#include <Common/logger_useful.h>
+#include <Common/typeid_cast.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+namespace
+{
+    constexpr bool debug_logging_enabled = true;
+    void logActionsDAG(const String & prefix, const ActionsDAGPtr & actions)
+    {
+        if constexpr (debug_logging_enabled)
+            LOG_DEBUG(&Poco::Logger::get("redundantDistinct"), "{}: {}", prefix, actions->dumpDAG());
+    }
+
+    void logDebug(String key, String value)
+    {
+        if constexpr (debug_logging_enabled)
+            LOG_DEBUG(&Poco::Logger::get("redundantDistinct"), "{} : {}", key, value);
+    }
+
+    using DistinctColumns = std::set<std::string_view>;
+    DistinctColumns getDistinctColumns(const DistinctStep * distinct)
+    {
+        /// find non-const columns in DISTINCT
+        const ColumnsWithTypeAndName & distinct_columns = distinct->getOutputStream().header.getColumnsWithTypeAndName();
+        std::set<std::string_view> non_const_columns;
+        const Names & column_names = distinct->getColumnNames();
+        for (const auto & column : distinct_columns)
+        {
+            if (!isColumnConst(*column.column) && find(cbegin(column_names), cend(column_names), column.name) != column_names.cend())
```

Comment:
I am worried that we are using a linear search here for every distinct column.
Maybe make a `std::unordered_set<std::string_view>`?